### PR TITLE
Improve UI for Deprecating Case Types in Data Dictionary

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

A new toggle button has been added to the case type list in the Data Dictionary page:
dd-new-button

Once the button has been clicked, a `load_deprecated_case_types=true` query param will be put in the URL and the button will switch to the following:
toggle-show-deprecated-dd

By default, deprecated case types will not be loaded/shown in the Data Dictionary.

## Technical Summary

Link to ticket [here](https://dimagi-dev.atlassian.net/browse/SC-3240).

A new UI button has been added to the sidebar of the Data Dictionary which gives the user the ability to toggle showing deprecated case types. This also has a potential performance benefit as domains with a lot of deprecated case types will now have faster loading times since deprecated case types are not loaded/processed by default.

## Feature Flag

None.

### Safety story

- Local testing.
- Unit test.

### Automated test coverage

Unit test has been created to ensure that deprecated case types are only added to the response JSON if the relevant query param has been given.

### QA Plan

No QA planned.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js [hunk 1]
```diff
@@ -420,6 +428,17 @@ hqDefine("data_dictionary/js/data_dictionary", [
             self.removefhirResourceType(false);
         };
 
+        self.toggleShowDeprecatedCaseTypes = function () {
+            self.showDeprecatedCaseTypes(!self.showDeprecatedCaseTypes());
+            const pageUrl = new URL(window.location.href);
+            if (self.showDeprecatedCaseTypes()) {
+                pageUrl.searchParams.append('load_deprecated_case_types', true);
```
Comment (SmittieC) [inline corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: is it necessary to add `load_deprecated_case_types` to the searchParams in the url if you're adding it as a param on line 252?

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js [hunk 2]
```diff
@@ -176,6 +176,14 @@ hqDefine("data_dictionary/js/data_dictionary", [
         self.showAll = ko.observable(false);
         self.availableDataTypes = typeChoices;
         self.fhirResourceTypes = ko.observableArray(fhirResourceTypes);
+
+        const params = new URLSearchParams(document.location.search);
+        self.showDeprecatedCaseTypes = ko.observable(params.get("load_deprecated_case_types") !== null);
+
+        // Elements with this class have a hidden class to hide them on page load. If we don't do this, then the elements
+        // will flash on the page for a bit while the KO bindings are being applied.
+        $(".deprecate-case-type").removeClass('hidden');
```
Comment (zandre-eng) [inline corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:185]: A better way to achieve this would be to put a `ko-template` class on a parent div and then it will stay hidden until `koApplyBindings()` is done on the element. 

To do it in this manner will require significant refactoring however, since the entire HTML template is binded to `dataDictionaryModel`. We would essentially need to break up this model and do separate bindings to chunks of the HTML template.

TL:DR - Given the refactoring that is required to achieve a more elegant outcome, I will leave it as such for now.

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: The reason I kept this solution was because I struggled to find a more elegant way of easily persisting the filter state after a page refresh. Ideally, it would be fantastic if we can remove having it in the URL and so I'm open to suggestions. :thinking:

Comment (SmittieC) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: I see. I'm not familiar with the interplay between the window.href url and the one used to retrieve the data from the BE, but my question boils down to: Does window.href have an influence on the URL used to retrieve the data from the BE?

Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: They are similar in that the `window.href` query param essentially gets passed to the BE request. When we click the toggle button, a query param gets put into `window.href` and the page refreshes. After the page refresh, we look if the query param exists and then pass that result to the BE request for fetching the JSON data.

Comment (SmittieC) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: I see, so it is ultimately needed. Thanks for the clarification

**Review by SmittieC — APPROVED** [review]
Looks good to me

File: corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: It unfortunately seems like it. Alternatively, we could simply store this filter value in the browser cache, but I don't know if that might just complicate things unnecessarily. :man_shrugging:

Comment (SmittieC) [reply corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:435]: Should be fine keeping it this way. If I'm not mistaken, Case List filters works the same way

**Comment by Charl1996** [comment]
I'm overall fine with the code, but it honestly kinda feels weird having the toggle in the sidebar. I don't have any better solutions, so maybe it's fine for now...

**Comment by zandre-eng** [comment]
@Charl1996 @orangejenny Agreed that the placement does feel a little unusual, however I think it would feel more out-of-place anywhere else since the right-side pane contains the details for a specific case type.

I originally wanted to go with a button in the sidebar. I ended up changing my mind as it didn't fit quite right, however I can be persuaded otherwise if others feel strongly about having a button instead.
Screenshot from 2024-01-11 09-26-59

> A couple of related suggestions:

Thanks for the suggestions! I like the idea of having it at the bottom, as it then groups all the interactive actions together.

**Review by orangejenny — APPROVED** [review]
approving last two commits - thanks for the updates!
